From 122a78c1fdfdda5f3cef38d3f5c31573b050e7d5 Mon Sep 17 00:00:00 2001
From: He Junyan <junyan.he@intel.com>
Date: Wed, 22 Feb 2023 16:49:38 +0800
Subject: [PATCH 103/127] va: Add modifier information to GstVaDmabufAllocator

We need the VA DMA allocator to create surfaces based on the
specified modifier value.
---
 .../gst-libs/gst/va/gstvaallocator.c            | 17 +++++++++++------
 .../gst-libs/gst/va/gstvaallocator.h            |  3 ++-
 .../gst-plugins-bad/gst-libs/gst/va/gstvapool.c |  5 ++++-
 3 files changed, 17 insertions(+), 8 deletions(-)

diff --git a/subprojects/gst-plugins-bad/gst-libs/gst/va/gstvaallocator.c b/subprojects/gst-plugins-bad/gst-libs/gst/va/gstvaallocator.c
index 922847e69c..42fe65eec0 100644
--- a/subprojects/gst-plugins-bad/gst-libs/gst/va/gstvaallocator.c
+++ b/subprojects/gst-plugins-bad/gst-libs/gst/va/gstvaallocator.c
@@ -273,6 +273,7 @@ struct _GstVaDmabufAllocator
   GstMemoryCopyFunction parent_copy;
 
   GstVideoInfo info;
+  guint64 modifier;
   guint usage_hint;
 
   GstVaSurfaceCopy *copy;
@@ -466,6 +467,7 @@ gst_va_dmabuf_allocator_init (GstVaDmabufAllocator * self)
   allocator->mem_map = gst_va_dmabuf_mem_map;
   self->parent_copy = allocator->mem_copy;
   allocator->mem_copy = gst_va_dmabuf_mem_copy;
+  self->modifier = DRM_FORMAT_MOD_INVALID;
 
   gst_va_memory_pool_init (&self->pool);
 }
@@ -694,8 +696,8 @@ gst_va_dmabuf_allocator_setup_buffer_full (GstAllocator * allocator,
 
   g_return_val_if_fail (GST_IS_VA_DMABUF_ALLOCATOR (allocator), FALSE);
 
-  if (!_va_dmabuf_create_and_export_surface (self->display, self->usage_hint,
-          DRM_FORMAT_MOD_INVALID, &self->info, &surface, &desc))
+  if (!_va_dmabuf_create_and_export_surface (self->display,
+          self->usage_hint, self->modifier, &self->info, &surface, &desc))
     return FALSE;
 
   buf = gst_va_buffer_surface_new (surface);
@@ -738,6 +740,7 @@ gst_va_dmabuf_allocator_setup_buffer_full (GstAllocator * allocator,
     gst_mini_object_set_qdata (GST_MINI_OBJECT (mem),
         gst_va_buffer_surface_quark (), buf, buffer_destroy);
 
+    g_assert (desc.objects[i].drm_format_modifier == self->modifier);
     *drm_mod = desc.objects[i].drm_format_modifier;
     gst_mini_object_set_qdata (GST_MINI_OBJECT (mem), gst_va_drm_mod_quark (),
         drm_mod, g_free);
@@ -960,10 +963,11 @@ gst_va_dmabuf_allocator_try (GstAllocator * allocator)
  * @allocator: a #GstAllocator
  * @info: a #GstVideoInfo
  * @usage_hint: VA usage hint
+ * @modifier: the underlying modifier
  *
- * Sets the configuration defined by @info and @usage_hint for
- * @allocator, and it tries the configuration, if @allocator has not
- * allocated memories yet.
+ * Sets the configuration defined by @info, @usage_hint and @modifier
+ * for @allocator, and it tries the configuration, if @allocator has
+ * not allocated memories yet.
  *
  * If @allocator has memory allocated already, and frame size and
  * format in @info are the same as currently configured in @allocator,
@@ -976,7 +980,7 @@ gst_va_dmabuf_allocator_try (GstAllocator * allocator)
  */
 gboolean
 gst_va_dmabuf_allocator_set_format (GstAllocator * allocator,
-    GstVideoInfo * info, guint usage_hint)
+    GstVideoInfo * info, guint usage_hint, guint64 modifier)
 {
   GstVaDmabufAllocator *self;
   gboolean ret;
@@ -999,6 +1003,7 @@ gst_va_dmabuf_allocator_set_format (GstAllocator * allocator,
 
   self->usage_hint = usage_hint;
   self->info = *info;
+  self->modifier = modifier;
 
   g_clear_pointer (&self->copy, gst_va_surface_copy_free);
 
diff --git a/subprojects/gst-plugins-bad/gst-libs/gst/va/gstvaallocator.h b/subprojects/gst-plugins-bad/gst-libs/gst/va/gstvaallocator.h
index cf1bcfc8a4..756b2078d1 100644
--- a/subprojects/gst-plugins-bad/gst-libs/gst/va/gstvaallocator.h
+++ b/subprojects/gst-plugins-bad/gst-libs/gst/va/gstvaallocator.h
@@ -50,7 +50,8 @@ void                  gst_va_dmabuf_allocator_flush       (GstAllocator * alloca
 GST_VA_API
 gboolean              gst_va_dmabuf_allocator_set_format  (GstAllocator * allocator,
                                                            GstVideoInfo * info,
-                                                           guint usage_hint);
+                                                           guint usage_hint,
+                                                           guint64 modifier);
 GST_VA_API
 gboolean              gst_va_dmabuf_allocator_get_format  (GstAllocator * allocator,
                                                            GstVideoInfo * info,
diff --git a/subprojects/gst-plugins-bad/gst-libs/gst/va/gstvapool.c b/subprojects/gst-plugins-bad/gst-libs/gst/va/gstvapool.c
index f5b77234bc..9e1db70b71 100644
--- a/subprojects/gst-plugins-bad/gst-libs/gst/va/gstvapool.c
+++ b/subprojects/gst-plugins-bad/gst-libs/gst/va/gstvapool.c
@@ -35,6 +35,9 @@
 #endif
 
 #include "gstvapool.h"
+#ifndef G_OS_WIN32
+#include <libdrm/drm_fourcc.h>
+#endif
 
 GST_DEBUG_CATEGORY_STATIC (gst_va_pool_debug);
 #define GST_CAT_DEFAULT gst_va_pool_debug
@@ -174,7 +177,7 @@ gst_va_pool_set_config (GstBufferPool * pool, GstStructure * config)
 
   if (GST_IS_VA_DMABUF_ALLOCATOR (allocator)) {
     if (!gst_va_dmabuf_allocator_set_format (allocator, &alloc_info,
-            usage_hint))
+            usage_hint, DRM_FORMAT_MOD_INVALID))
       goto failed_allocator;
   } else if (GST_IS_VA_ALLOCATOR (allocator)) {
     if (!gst_va_allocator_set_format (allocator, &alloc_info, usage_hint,
-- 
2.25.1

